core/src/main/scala/com/pingcap/tispark/utils/WriteUtil.scala

```diff
@@ -114,116 +118,152 @@ object WriteUtil {
    * Generate Record that will be removed
    * key: tableId + handle
    * value: empty
+   *
    * @param rdd
    * @param tableId
    * @return
    */
-  def generateRecordKVToDelete(rdd: RDD[WrappedRow], tableId: Long): RDD[WrappedEncodedRow] = {
+  def generateRecordKVRDDToDelete(rdd: RDD[WrappedRow], tableId: Long): RDD[WrappedEncodedRow] = {
     rdd.map { wrappedRow =>
       {
-        val (encodedKey, encodedValue) = (
-          new SerializableKey(RowKey.toRowKey(tableId, wrappedRow.handle).getBytes),
-          new Array[Byte](0))
-        WrappedEncodedRow(
-          wrappedRow.row,
-          wrappedRow.handle,
-          encodedKey,
-          encodedValue,
-          isIndex = false,
-          -1,
-          remove = true)
+        generateRecordKVToDelete(wrappedRow, tableId)
       }
     }
   }
 
+  def generateRecordKVToDelete(wrappedRow: WrappedRow, tableId: Long): WrappedEncodedRow = {
+    val (encodedKey, encodedValue) = (
+      new SerializableKey(RowKey.toRowKey(tableId, wrappedRow.handle).getBytes),
+      new Array[Byte](0))
+    WrappedEncodedRow(
+      wrappedRow.row,
+      wrappedRow.handle,
+      encodedKey,
+      encodedValue,
+      isIndex = false,
+      -1,
+      remove = true)
+  }
+
   /**
    * use all indices to generate Index kv.
    * For isCommonHandle, we exclude primary key for it has been built by record
    * For isPkHandle, we don't do this because primary key is not included in indices
+   *
    * @param rdd
    * @param remove
-   * @param tiTableInfo
-   * @return  Map[Long, RDD[WrappedEncodedRow], The key of map is indexId
+   * @param tiTable
+   * @return Map[Long, RDD[WrappedEncodedRow], The key of map is indexId
    */
-  def generateIndexKVs(
+  def generateIndexKVRDDs(
       rdd: RDD[WrappedRow],
-      tiTableInfo: TiTableInfo,
+      tiTable: TableCommon,
       remove: Boolean): Map[Long, RDD[WrappedEncodedRow]] = {
-    tiTableInfo.getIndices.asScala.flatMap { index =>
-      if (tiTableInfo.isCommonHandle && index.isPrimary) {
+    val tableInfo = tiTable.getTableInfo
+    tableInfo.getIndices.asScala.flatMap { index =>
+      if (tableInfo.isCommonHandle && index.isPrimary) {
         None
       } else {
-        Some((index.getId, generateIndexRDD(rdd, index, tiTableInfo, remove)))
+        Some((index.getId, generateIndexRDD(rdd, index, tiTable, remove)))
       }
     }.toMap
   }
 
+  def generateIndexKVs(
+      rdd: WrappedRow,
+      tiTable: TableCommon,
+      remove: Boolean): mutable.Map[Long, mutable.Set[WrappedEncodedRow]] = {
+    val tableInfo = tiTable.getTableInfo
+    listPair2multimap(tableInfo.getIndices.asScala.flatMap { index =>
+      if (tableInfo.isCommonHandle && index.isPrimary) {
+        None
+      } else {
+        Some((index.getId, generateIndex(rdd, index, tiTable, remove)))
+      }
+    }.toList)
+  }
+
   /**
    * mix the results that are produced by method generateIndexKVs
+   *
```

Comment:
Should @param `tiTableInfo` be `TableCommon`?

Reply:
done